Auto FBX

infernoplus opened the pull request:
Automatically finds and converts FBX files to JMS or ASS when you run commands like collision, physics, or structure.
Also, if you place an fbx file in the root of a directory with the extension ".all.fbx" it will be used for render, collision, and physics.
EX: "data/objects/vehicles/test/test.fbx.all" 
This should make the workflow a lot smoother.

Let me know if I should adjust anything. A bit messy.

Adding support for H1/H2 should be a real easy copy paste job but I need to know the folder structure and commands for each game.

num0005 commented on Launcher/ToolkitInterface/H1AToolkit.cs:
```diff
@@ -67,7 +67,7 @@ protected override string sapienWindowClass
         /// <param name="h2SelectionLogic"></param>
         /// <returns></returns>
-        public override async Task ImportModel(string path, ModelCompile importType, bool phantomFix, bool h2SelectionLogic, bool renderPRT, bool FPAnim, string characterFPPath, string weaponFPPath, bool accurateRender, bool verboseAnim, bool uncompressedAnim, bool skyRender, bool resetCompression)
+        public override async Task ImportModel(string path, ModelCompile importType, bool phantomFix, bool h2SelectionLogic, bool renderPRT, bool FPAnim, string characterFPPath, string weaponFPPath, bool accurateRender, bool verboseAnim, bool uncompressedAnim, bool skyRender, bool resetCompression, bool autoFBX)
```
Note to self: Open an issue for reworking how arguments are passed to some of these functions.

num0005 commented on Launcher/ToolkitInterface/H3Toolkit.cs:
```diff
@@ -32,8 +33,21 @@ override public async Task ImportUnicodeStrings(string path)
+        private async Task StructureAutoFBX(string data_file)
         {
+            /* Convert Structure */
+            string inPath = BaseDirectory + "\\data\\" + data_file.ToLower().Replace(".ass", ".fbx"); // Jank code, should probably make this safer
```
This should use ` GetDataDirectory` https://github.com/num0005/Osoyoos-Launcher/blob/b470fcc3787d752e418dbd072306f772beb2d2c7/Launcher/ToolkitInterface/H1AToolkit.cs#L45
H3 might get data directory overrides in a future build.

num0005 commented on Launcher/ToolkitInterface/H3Toolkit.cs:
```diff
@@ -32,8 +33,21 @@ override public async Task ImportUnicodeStrings(string path)
         {
+            /* Convert Structure */
+            string inPath = BaseDirectory + "\\data\\" + data_file.ToLower().Replace(".ass", ".fbx"); // Jank code, should probably make this safer
+            if (File.Exists(inPath))
```
Should this log or display an error message if the path is missing?

num0005 commented on Launcher/ToolkitInterface/H3Toolkit.cs:
```diff
@@ -32,8 +33,21 @@ override public async Task ImportUnicodeStrings(string path)
+            string inPath = BaseDirectory + "\\data\\" + data_file.ToLower().Replace(".ass", ".fbx"); // Jank code, should probably make this safer
+            if (File.Exists(inPath))
+            {
+                string outPath = BaseDirectory + "\\data\\" + data_file;
```
ditto

num0005 commented on Launcher/ToolkitInterface/H3Toolkit.cs:
```diff
@@ -174,13 +188,133 @@ public override async Task BuildLightmap(string scenario, string bsp, LightmapAr
 
         }
 
+        // Some nonsense, probably should be an enum actually
+        private class ImportTypeInfo
+        {
+            public ModelCompile importType;
+            public string folderName, commandOption;
+            public ImportTypeInfo(ModelCompile a, string b, string c)
+            {
+                importType = a;
+                folderName = b;
+                commandOption = c;
+            }
+        }
```
Maybe make this a record?
https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/tutorials/records

num0005 commented on Launcher/ToolkitInterface/H3Toolkit.cs:
```diff
@@ -174,13 +188,133 @@ public override async Task BuildLightmap(string scenario, string bsp, LightmapAr
+        {
+            /* Check for x.all.fbx in root directory. */
+            List<FileInfo> alls = new List<FileInfo>();
+            string rootDir = BaseDirectory + "\\data\\" + path;
```
same data directory thing.

num0005 commented on Launcher/ToolkitInterface/H3Toolkit.cs:
```diff
@@ -174,13 +188,133 @@ public override async Task BuildLightmap(string scenario, string bsp, LightmapAr
+            {
+                foreach (FileInfo all in alls)
+                {
+                    string outPath = rootDir + "\\" + (all.Name.ToLower().Replace(".all.fbx", ".jms"));
```
No need to change this but if this needs the ToLower please add a comment explaining why so no-one removes it.

num0005 commented on Launcher/ToolkitInterface/H3Toolkit.cs:
```diff
@@ -174,13 +188,133 @@ public override async Task BuildLightmap(string scenario, string bsp, LightmapAr
+            List<FileInfo> alls = new List<FileInfo>();
+            string rootDir = BaseDirectory + "\\data\\" + path;
+
+            string[] rootFilePaths = Directory.GetFiles(rootDir);
+            foreach (string f in rootFilePaths)
+            {
+                FileInfo fi = new FileInfo(f);
+                if (fi.Name.ToLower().EndsWith(".all.fbx"))
+                {
+                    alls.Add(fi);
+                }
+            }
```
GetFiles supports a search pattern
https://docs.microsoft.com/en-us/dotnet/api/system.io.directory.getfiles?view=net-6.0#System_IO_Directory_GetFiles_System_String_System_String_
Should simplify the code.

num0005 commented on Launcher/ToolkitInterface/H3Toolkit.cs:
```diff
@@ -174,13 +188,133 @@ public override async Task BuildLightmap(string scenario, string bsp, LightmapAr
+                {
+                    string typeDir = rootDir + "\\" + importTypeInfo.folderName;
+
+                    /* Copy x.all.fbx output if it exists */
```
Is this why the output path above was normalized to lowercase? Why can't the file be outputted directly to the final directory instead of being copied over afterwards?

num0005 reviewed (changes requested):
Overall it's a nice feature and would be great to support it for all toolkits.

I think you could move the implementation to a separate static class or something like that and use the interfaces setup for FBX to make it work for all games. Maybe there is a more natural way to do that in C# just the best thing I can think of right now. Really wish C# had multiple inheritance.

infernoplus commented:
> I think you could move the implementation to a separate static class or something like that and use the interfaces setup for FBX to make it work for all games. Maybe there is a more natural way to do that in C# just the best thing I can think of right now. Really wish C# had multiple inheritance.

Moved it to it's own class and added support for all games. 
Someone else should test the support for the other games because I don't have them installed. I just looked at the wiki for differences in the folder names and commands. Hopefully I got it all right haha.

num0005 reviewed (approved):
The code seems fine, someone just needs to test it. @General-101 when you have time could you take a look at whatever it actually works?

infernoplus commented:
Just found an odd issue with tool.exe in Halo 3 at least that I need to address. Made a post here about it. Gonna write a util to solve this. If you get a chance can either of you confirm this bug exists in Halo 1/2 tool?

https://steamcommunity.com/app/976730/discussions/3/3120423877859517091/?ctp=5#c3200370471664585592

num0005 commented:
That's an interesting bug, will look into it.

num0005 reviewed (changes requested):
The UI needs to be setup to support the AutoFBX for H1A, right now there is no way to enable it. Going to test it to see if works for H2C in a bit but at least the UI works.

infernoplus commented:
Went down some rabbit holes and wrote a rather uh... interesting... class that can load an FBX, strip parts out of it, and write it back to file.

It does seem to solve the bug, though I was initially off. It's the physics that are getting copied into the render JMS, not the collision. No idea why.

I'll commit changes with this new stuff soonish.

num0005 commented:
> As for the H1A UI, I actually don't know how to add that. What do?

Not actually sure right now that code is a bit convoluted and could use a cleanup, I think it would be fine to handle adding that in a different PR.

infernoplus commented:
Heh, fixed the function names. Trying to find an api to read & write FBX left me pretty tilted.

As for writing binary vs ascii, I found that the binary doesn't always play nice on export so I'm using ascii. The ascii file is temporary though and is automatically deleted after compiling. It only exists on the drive for a few moments.

Reading binary is fine though, writing just doesn't work for whatever reason.

And on the subject of that, I have an exception catch on the fbx strip code, but no actual logging. How should I handle logging for warnings/errors?

num0005 commented:
I don't think logging is setup yet but just use `Trace` https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.trace.traceerror?view=net-6.0
When it gets setup it will use it too.

num0005 reviewed (approved):
LGTM
